src/lib/sw_engine/tvgSwRaster.cpp

```diff
@@ -325,17 +345,24 @@ static bool _rasterTranslucentImageRle(SwSurface* surface, const SwRleData* rle,
 static bool _rasterTranslucentImageRle(SwSurface* surface, const SwRleData* rle, uint32_t *img, uint32_t w, uint32_t h, uint32_t opacity, const Matrix* invTransform)
 {
     auto span = rle->spans;
-
+    bool isDownScaling = (invTransform->e11 * invTransform->e11) + (invTransform->e21 * invTransform->e21) > 1 ? true : false;
```

Comment:
yes, I also thought about using a bool :)

and like below? 
I know that a huge amount of the code is duplicated. I'm not sure whether it's worth of it. 
what do you think? 

```
if (isDownScaling) {
    for (uint32_t i = 0; i < rle->size; ++i, ++span) {
        auto ey1 = span->y * invTransform->e12 + invTransform->e13;
        auto ey2 = span->y * invTransform->e22 + invTransform->e23;
        auto dst = &surface->buffer[span->y * surface->stride + span->x];
        auto alpha = ALPHA_MULTIPLY(span->coverage, opacity);
        for (uint32_t x = 0; x < span->len; ++x, ++dst) {
            auto rX = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e11 + ey1));
            auto rY = static_cast<uint32_t>(roundf((span->x + x) * invTransform->e21 + ey2));
            if (rX >= w || rY >= h) continue;
            auto src = ALPHA_BLEND(img[rY * w + rX], alpha);     //TODO: need to use image's stride
            *dst = src + ALPHA_BLEND(*dst, 255 - surface->blender.alpha(src));
        }
    }
} else {
    for (uint32_t i = 0; i < rle->size; ++i, ++span) {
        auto ey1 = span->y * invTransform->e12 + invTransform->e13;
        auto ey2 = span->y * invTransform->e22 + invTransform->e23;
        auto dst = &surface->buffer[span->y * surface->stride + span->x];
        auto alpha = ALPHA_MULTIPLY(span->coverage, opacity);
        for (uint32_t x = 0; x < span->len; ++x, ++dst) {
            auto fX = (span->x + x) * invTransform->e11 + ey1;
            auto fY = (span->x + x) * invTransform->e21 + ey2;
            auto rX = static_cast<uint32_t>(fX);
            auto rY = static_cast<uint32_t>(fY);
            if (rX >= w || rY >= h) continue;
            uint32_t src;
            if (rX == w - 1 || rY == h - 1) {
                src = ALPHA_BLEND(img[rY * w + rX], alpha);     //TODO: need to use image's stride
            } else {
                src = ALPHA_BLEND(_applyBilinearInterpolation(img, w, h, fX, fY), alpha);     //TODO: need to use image's stride
            }
            *dst = src + ALPHA_BLEND(*dst, 255 - surface->blender.alpha(src));
        }
    }
}
```

also please notice, that in the `else` block I removed a floor function when rX and rY are calculated.

Reply:
Thank you for your review. I have split the raster function as per your comment. So I added the appropriate functions.